data-prepper-plugins/newline-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/newline/NewlineDelimitedOutputCodec.java

```diff
@@ -4,42 +4,89 @@
  */
 package org.opensearch.dataprepper.plugins.codec.newline;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
 import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
 import org.opensearch.dataprepper.model.codec.OutputCodec;
 import org.opensearch.dataprepper.model.event.Event;
 
 import java.io.IOException;
 import java.io.OutputStream;
+import java.util.Map;
+import java.util.Objects;
 
 /**
  * An implementation of {@link OutputCodec} which deserializes Data-Prepper events
  * and writes them to Output Stream as ND-JSON data
  */
-@DataPrepperPlugin(name = "newline", pluginType = OutputCodec.class)
+@DataPrepperPlugin(name = "newline", pluginType = OutputCodec.class, pluginConfigurationType = NewlineDelimitedOutputConfig.class)
 public class NewlineDelimitedOutputCodec implements OutputCodec {
+    private static final String NDJSON = "ndjson";
+    private static final String MESSAGE_FIELD_NAME = "message";
+    private final String headerDestination;
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+    private final NewlineDelimitedOutputConfig config;
+
+    @DataPrepperPluginConstructor
+    public NewlineDelimitedOutputCodec(final NewlineDelimitedOutputConfig config) {
+        Objects.requireNonNull(config);
+        this.config = config;
+        headerDestination = config.getHeaderDestination();
+    }
 
     @Override
     public void start(final OutputStream outputStream) throws IOException {
```

Comment:
I wasn't involved in the design reviews or previous CR's but this interface looks wrong to me. OutputCodec should either write a batch of events to an OutputFile or a batch of events to a ByteArrayOutputStream. In the case of S3 you would want to write to an S3OutputFile. This will make more sense when you write the ParquetOutputCodec. S3 sink then needs to be responsible for file truncating logic. Maybe the customer needs to specify a max watermark or max file size that dictates when files are truncated.

Reply:
In codec we are converting and appending record in same outputstream referred by buffer. Threshold check/flushing the same buffer taken care in the S3SinkService

Reply:
Yes.! need to finalize Architecture for Parquet.